src/OrbitBase/include/OrbitBase/Typedef.h

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_TYPEDEF_H_
+#define ORBIT_BASE_TYPEDEF_H_
+
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace orbit_base {
+
+template <typename T, typename U>
+using EnableIfUConvertibleToT = std::enable_if_t<std::is_convertible_v<U, T>>;
+
+template <typename T, typename U>
+using EnableIfUIsNotConvertibleToT = std::enable_if_t<!std::is_convertible_v<U, T>>;
+
+// Strong typedef
+template <typename Tag, typename T>
+class Typedef {
+ public:
+  template <typename OtherTag, typename U>
+  friend class Typedef;
+
+  [[nodiscard]] constexpr const T* operator->() const { return &value_; }
+  [[nodiscard]] constexpr T* operator->() { return &value_; }
+
+  [[nodiscard]] constexpr const T& operator*() const& { return value_; }
+  [[nodiscard]] constexpr T&& operator*() && { return std::move(value_); }
+  [[nodiscard]] constexpr const T&& operator*() const&& { return std::move(value_); }
+  [[nodiscard]] constexpr T& operator*() & { return value_; }
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr explicit Typedef(U&& value) : value_(std::forward<U>(value)) {}
+
+  template <typename... Args>
+  constexpr explicit Typedef(std::in_place_t, Args&&... args)
+      : value_(T(std::forward<T>(args)...)) {}
+
+  constexpr Typedef(Typedef&& other) = default;
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr Typedef(const Typedef<Tag, U>& other) : value_(other.value_) {}
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr Typedef(Typedef<Tag, U>&& other) : value_(std::move(other.value_)) {}
+
+  template <typename U, typename = EnableIfUIsNotConvertibleToT<T, U>, typename = void>
+  constexpr explicit Typedef(const Typedef<Tag, U>& other) : value_(other.value_) {}
+
+  template <typename U, typename = EnableIfUIsNotConvertibleToT<T, U>, typename = void>
+  constexpr explicit Typedef(Typedef<Tag, U>&& other) : value_(std::move(other.value_)) {}
```

Comment:
I don't understand the additional `typename = void` parameter. You said without it it didn't work. Have you figured out why? I would love to understand if that's really fixing a potential problem or just silencing the symptom.

Reply:
> You said without it it didn't work. Have you figured out why? 
I think, the compiler generates an error at the first phase of name lookup. There it does see a redefinition of constructor that differs only by constexpr (and you can't overload by constexpr) and by the default template value, which starts to matter at the second phase.